onnxruntime/core/graph/graph_utils.cc

```diff
@@ -399,9 +399,24 @@ bool IsGraphInput(const Graph& graph, const NodeArg* input) {
   return std::find(graph_inputs.begin(), graph_inputs.end(), input) != graph_inputs.end();
 }
 
+bool IsConstantInitializer(const Graph& graph, const std::string& initializer_name) {
+  bool constant = true;
+
+  if (graph.CanOverrideInitializer()) {
```

Comment:
>graph.CanOverrideInitializer() [](start = 6, length = 30)

For initializers in subgraph, do we allow them to be overridden?

Reply:
As currently implemented it's all or none. The spec says the inputs to the control flow node should match the subgraph inputs, however a model with IR version of 3 would have inputs for initializers. Due to that we first try to match all inputs (including initializers), and fallback to matching just required inputs. There's no way to partially match optional inputs as there's no information on which optional inputs are being provided.

We could possibly make it stricter if IR version is >= 4 and require exact match of all inputs.

---
In reply to: [299590945](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299590945) [](ancestors = 299590945)

Reply:
If I understood correctly, seems we can add condition in graph.CanOverrideInitializer() to make subgraphs always return false.

---
In reply to: [299710986](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299710986) [](ancestors = 299710986,299590945)

Reply:
Glad to see the recursive search. Do we need a testcase with subgraph for it?

---
In reply to: [299745451](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299745451) [](ancestors = 299745451,299710986,299590945)

Reply:
@kkaranasos Do you know if any of the optimizer or model tests will cover this?

Reply:
Will follow up with kkaranasos to make sure this gets covered by tests.